### PR TITLE
deployment: add helm charts, configuration and openshift docs

### PR DESCRIPTION
**Pull request by ppanero**

### CODE REVIEW

**Comment by fenekku** [comment]
Let's change the link in `docs/index.md` from `[> Deployment Guides](deployment/openshift.md)` to the new `[> Deployment Guides](deployment/index.md)`

File: docs/deployment/index.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+$ helm repo add helm-invenio https://inveniosoftware.github.io/helm-invenio/
+$ helm repo update
+$ helm search invenio
+
+NAME                   	CHART VERSION	APP VERSION	DESCRIPTION
+helm-invenio/invenio	0.2.0        	1.16.0     	Open Source framework for large-scale digital repositories
+helm-invenio/invenio	0.1.0        	1.16.0     	Open Source framework for large-scale digital repositories
+```
+
+You can also install by cloning from GitHub by cloning the repository:
```
Comment (fenekku) [inline docs/deployment/index.md]:
```suggestion
You can also install by cloning the repository from GitHub:
```

File: docs/deployment/index.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+```
+$ git clone https://github.com/inveniosoftware/helm-invenio.git
+$ cd helm-invenio/
+```
+
+Then you will need, to reference the `./invenio` folder rather than the chart name (`helm-invenio/invenio`).
```
Comment (fenekku) [inline docs/deployment/index.md]:
```suggestion
In this case, you will need to reference the `./invenio` folder rather than the chart name (`helm-invenio/invenio`).
```

File: docs/deployment/index.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+## Supported Platforms
+
+!!! warning "Only compatible with OpenShift"
+    Pleas note that currently these Helm charts are only compatible with OpenShift.
```
Comment (fenekku) [inline docs/deployment/index.md]:
```suggestion
    Please note that currently these Helm charts are only compatible with OpenShift.
```

File: docs/deployment/openshift.md [hunk 4]
```diff
@@ -1,3 +1,162 @@
...
\ No newline at end of file
+## Pre-Requirements
+
+- [Global deployment pre-requirements](index.md#pre-requirements)
+- [OpenShift CLI](https://docs.openshift.com/container-platform/4.3/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) version 3.11+
+
+
+## Deploying InvenioRDM
+
+First of all login and select the right project in your OpenShift cluster:
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
First of all, login and select the right project in your OpenShift cluster:
```

File: docs/deployment/openshift.md [hunk 5]
```diff
@@ -1,3 +1,162 @@
...
+
+```console
+$ oc login <your.openshift.cluster>
+$ oc project invenio
+```
+
+### Secrets
+
+Before deploying in need to provide the credentials so the application can access the different services:
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
Before deploying, you need to provide the credentials so the application can access the different services. Replace the credentials and configuration below with your own:
```

File: docs/deployment/openshift.md [hunk 6]
```diff
@@ -1,3 +1,162 @@
...
+$ oc create secret generic \
+  --from-literal="RABBITMQ_DEFAULT_PASS=$RABBITMQ_DEFAULT_PASS" \
+  --from-literal="CELERY_BROKER_URL=amqp://guest:$RABBITMQ_DEFAULT_PASS@mq:5672/" \
+  mq-secrets
+secret "mq-secrets" created
+```
+
+**Elasticsearch secrets:**
+
+!!! info "Elasticaserch variables"
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
!!! info "Elasticsearch variables"
```

File: docs/deployment/openshift.md [hunk 7]
```diff
@@ -1,3 +1,162 @@
...
+
+``` console
+$ export INVENIO_SEARCH_ELASTIC_HOSTS="[{'host': 'localhost', 'timeout': 30, 'port': 9200, 'use_ssl': True, 'http_auth':('USERNAME_CHANGEME', 'PASSWORD_CHANGEME')}]"
+$ oc create secret generic \
+  --from-literal="INVENIO_SEARCH_ELASTIC_HOSTS=$INVENIO_SEARCH_ELASTIC_HOSTS" \
+  elasticsearch-secrets
+```
+
+!!! info "Extra configuration is possible"
+    Note that you might need to add extra configuration to the elasticsearch hosts, sucha as vertificate verification (`verify_certs`), prefixing (`url_prefix`) and more.
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
    Note that you might need to add extra configuration to the Elasticsearch hosts, such as certificate verification (`verify_certs`), prefixing (`url_prefix`) and more.
```

File: docs/deployment/openshift.md [hunk 8]
```diff
@@ -1,3 +1,162 @@
...
+
+### Install InvenioRDM
+
+Before installing you need to configure two things, the rest are optional and you can read more about it [here](configuration.md):
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
Before installing you need to configure two things in your `values.yml` file. The rest are optional and you can read more about them [here](configuration.md):
```

File: docs/deployment/openshift.md [hunk 9]
```diff
@@ -1,3 +1,162 @@
...
+
+- Your host in a `values.yaml` file.
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
- The host.
```

File: docs/deployment/openshift.md [hunk 10]
```diff
@@ -1,3 +1,162 @@
...
+host: yourhost.localhost
+
+web:
+  image: your/invenio-image
+
+worker:
+  image: your/invenio-image
+```
+
+The next step is the installation itself, with your own configuration in the `values.yaml`. If you added the repository you can install it by using the chart name and the desired version:
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
The next step is the installation itself, with your own configuration in the `values.yaml`. If you added the repository, you can install it by using the chart name and the desired version:
```

File: docs/deployment/openshift.md [hunk 11]
```diff
@@ -1,3 +1,162 @@
...
+In some cases you might want to run jobs, for example to populate the instance with records.
+
+``` console
+$ oc process -f job.yml --param JOB_NAME='demo-data-1' \
+  --param JOB_COMMAND='invenio rdm-records demo' | oc create -f -
+```
+
+**Cron job**
+
+Now imagine you might have some bulk record creation that needs to be indexed, or any other task that you have to do every certain period of time.
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
Now, imagine you have some bulk record indexing, or any other task that you have to do periodically.
```

File: docs/deployment/openshift.md [hunk 12]
```diff
@@ -1,3 +1,162 @@
...
+For that you can define cronjobs:
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
For that, you can define cronjobs:
```

File: docs/deployment/openshift.md [hunk 13]
```diff
@@ -1,3 +1,162 @@
...
+
+``` console
+$ oc process -f cronjob.yml --param JOB_NAME=index-run \
+  --param JOB_COMMAND=invenio index run -d | oc create -f -
+```
+
+### Upgrade your instance
+
+If you have performed some changes to your instance (e.g. configuration) or you want to upgrade the version of the charts, you can do so with
+the `upgrade` command of `helm`, note that you still need to disable the openapi validation which is only supoprted after version 3.1.2:
```
Comment (fenekku) [inline docs/deployment/openshift.md]:
```suggestion
the `upgrade` command of `helm`. Note that you still need to disable the openapi validation which is only supported after version 3.1.2:
```

File: mkdocs.yml [hunk 14]
```diff
@@ -43,6 +43,9 @@ nav:
     - Add your extensions: 'extensions/custom.md'
     - S3 Storage: 'extensions/s3.md'
   - Deploy:
+    - How can I deploy it?: 'deployment/index.md'
+    - System components: 'deployment/services.md'
+    - Configuration:  'deployment/configuration.md'
     - Kubernetes: 'deployment/kubernetes.md'
     - OpenShift: 'deployment/openshift.md'
```
Comment (fenekku) [inline mkdocs.yml]: How about the following order to get to the "higher-level" pages first (how to deploy on platforms) and then the lower-level pages after (the configuration definitions)? Like we do for most packages, where we place the API at the end. Especially since System components is empty right now.
 
```suggestion
    - How can I deploy it?: 'deployment/index.md'
    - OpenShift: 'deployment/openshift.md'
    - Kubernetes: 'deployment/kubernetes.md'
    - System components: 'deployment/services.md'
    - Configuration:  'deployment/configuration.md'
```

File: docs/deployment/configuration.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
+# Configuration
+
+This section explain the current configuration options that are available for the different components that are deployed but the Helm charts.
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
This section explains the configuration options that are available for the different components deployed by the InvenioRDM Helm chart.
```

(it's one chart for InvenioRDM, right?)

File: docs/deployment/configuration.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+## Global
+
+The is only one mandatory configuration, which is the host name, for example:
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
There is only one mandatory configuration: the host name.
```

File: docs/deployment/configuration.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+```yaml
+host: your-rdm-instance.com
+```
+
+Moreover, the services can be deployed along, note that it is recommended to deploy separatelly Elasticsearch and PostgreSQL for a production deployment.
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
Moreover, the services can be deployed along. Note that it is recommended to deploy Elasticsearch and PostgreSQL separately for a production deployment.
```

File: docs/deployment/configuration.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+  inside_cluster: false
+```
+
+!!! info "inside_cluster availability
+    Note that the `inside_cluster` variable is supported for `redis`, `rabbitmq`, `elasticsearch`, `postgresql` and `haproxy`. The rest of the components
+    are mandatory.
+
+## HAProxy
+
+You can change the number of connections allowed by the haproxy with the `maxconn` variable:
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
You can change the number of connections allowed by haproxy with the `maxconn` variable:
```

File: docs/deployment/configuration.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+!!! warning "Only one parent element"
+    By default the HAProxy is enabled, `inside_cluster` has the value `true`, nonetheless if you decide to set it, you only need
+    to specify once the `haproxy` parent in the yaml file. Otherwise the last one will override the previous. It should be something like:
+    ``` yaml
+    haproxy:
+        inside_cluster: true
+        maxconn: 100
+    ```
```
Comment (fenekku) [inline docs/deployment/configuration.md]: I don't think I understand this part. Could you rephrase it?  :smiley_cat: (haproxy is enabled by default... so why do I need to specify it? What do I "set" and what is the "last one" that "will override the previous"?)

File: docs/deployment/configuration.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+## Nginx
+
+The charts allow you to configure the amount of connections per nginx node (replica) and the amount of nodes:
```
Comment (fenekku) [inline docs/deployment/configuration.md]: 1 chart <==> 1 package . "chart" is singular or is it "charts" because multiple chart are combined together?

File: docs/deployment/configuration.md [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+```
+nginx:
+  max_conns: 100
+  replicas: 2
+```
+
+## Web nodes
+
+The web nodes host the WSGI application, in order to be scalable you can configure the number of "nodes", called replicas, how many processes do each node run and with how many threads per process. The only mandatory parameter is the docker image (`image`) that should get as value the url where to pull the image from.
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
The web nodes host the WSGI application. In order to be scalable, you can configure the number of "nodes", called replicas, with how many processes each node runs and with how many threads per process. The only mandatory parameter is the docker image (`image`) that should get as value the url where to pull the image from.
```

File: docs/deployment/configuration.md [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+In addition, you can add automatic scaling, by setting minimum and maximum replicas and the threshold of cpu usage in which a new node should be spawned. For example, with a threshold of 65%, it meand that when the average CPU utilization of the nodes reaches 65% a new node will de spawned, till it reaches the setted maximum:
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
In addition, you can add automatic scaling by setting minimum (`min_web_replicas`) and maximum (`max_web_replicas`) replicas, and the cpu usage threshold in percentage (`scaler_cpu_utilization`) that spawns a new node. For example, with a `scaler_cpu_utilization` value of 65, it means that when the average CPU utilization of the nodes reaches 65%, a new node will be spawned. This process will repeat itself until the maximum number of replicas has been reached:
```

File: docs/deployment/configuration.md [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+    enabled: false
+    # Scale when CPU usage gets to
+    scaler_cpu_utilization: 65
+    max_web_replicas: 10
+    min_web_replicas: 2
+```
+
+## Worker nodes
+
+Finally, the worker nodes. By default they are enabled, but you can cancel their deployment by setting `enabled` to `false`. If enabled, in the same fashion than the web nodes
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
Finally, the worker nodes. By default they are enabled, but you can cancel their deployment by setting `enabled` to `false`. If enabled, they require an `image` like the web nodes.
```

File: docs/deployment/configuration.md [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+they require an `image`.
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
```

File: docs/deployment/configuration.md [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+
+In addition, you can configure how many worker nodes (replicas) will be deployed, which the application they will run, with which concurrency level and the logging level.
```
Comment (fenekku) [inline docs/deployment/configuration.md]:
```suggestion
In addition, you can configure the number of worker nodes (replicas) to be deployed, the application they will run, their concurrency level and their logging level.
```

File: docs/deployment/configuration.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (ppanero) [reply docs/deployment/configuration.md]: Nope, it is in a block of code just below :)

**Comment by ppanero** [comment]
@fenekku change done :) thanks for the review, can it be merged?

File: docs/deployment/index.md [hunk 26]
```diff
@@ -0,0 +1,60 @@
+# How can I deploy InvenioRDM?
+
+You can deploy InvenioRDM in several ways. You can install it in your [local computer](../develop/index.md) to have more easily customizable environment, otherwise you can deploy a [containerized environment](../preview/index.md) that demonstrates the setup of all components runnning in docker containers. In this section it is explained how to deploy in a closer-to-production manner.
+
+!!! warning "Do not deploy as-is in production"
+    Please note that it is mentioned as "closer-to-production", this is because even if the designed architecture can scale and withstand the load of a production service (It has been tested to stand peaks of up to 180 requests/s), the security configurations might not be enough and you should review it. In addition, it can deploy the extra services (Elasticsearch and PostgreSQL) along with them, however this ones are not configured with redundancy and persistance.
```
[the remark this answers is not in the document]
Comment (ppanero) [reply docs/deployment/index.md]: I see your point. IMHO ES and DB shouldn't be deployed along with this, but for demo purposes we have to. In addition, is not about the volumes or not, is the fact that if you "helm uninstall" it all gets wiped out, so does the volume, with an external DB you assure that doesn't happen (e.g. the demo that I did on the telecon uninstall --> install and data was still there.

I think this documentation will be more clear when in the future we have the "services" properly described. There will be a more detailed needs. This is intended as a recipe, that will become able to do prod deployment, but atm we are in beta, so lets see how it evolves (along with the docs ofc). There are many corners still to be polished :)

P.S. I have made some changes and tried to adapt it better.

File: docs/deployment/configuration.md (hunk 15 above)
Comment (ppanero) [reply docs/deployment/configuration.md]: not really, is just the image that is configurable. Once we set it up we will put it in the docs, but this charts suit all Invenio instances.

File: docs/deployment/configuration.md (hunk 19 above)
Comment (ppanero) [reply docs/deployment/configuration.md]: It is a case that should not happen. If it is enabled you only, might, need to change the maxcoon. Otherwise just disable it. My point was more of a yaml syntax point. That it should be:

```yaml
haproxy:
    inside_cluster: true
    maxconn: 100
```

because if you do:

```yaml
haproxy:
    inside_cluster: true

haproxy:
    maxconn: 100
```

only the second one applies, but in this case doesn't really make sense since it is active by default. So I removed the paragraph.

File: docs/deployment/configuration.md (hunk 20 above)
Comment (ppanero) [reply docs/deployment/configuration.md]: I have seen both in the literature, I have not seen a naming convention. Will look for it.

File: docs/deployment/index.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (fenekku) [reply docs/deployment/index.md]: > "helm uninstall" it all gets wiped out

Oh man, I don't want to be the person deploying things :sweat_smile: 

> more clear when in the future we have the "services" properly described. There will be a more detailed needs

Makes sense! Overall I am sure with time, the details will be added and more helpful tips will be provided.

File: docs/deployment/configuration.md (hunk 15 above)
Comment (fenekku) [reply docs/deployment/configuration.md]: To be clear, in the updated configuration.md, what do "these Helm charts" refer to? Is the invenio chart an amalgamation of charts (or maybe what I am saying makes no sense in Kubernetes world?)

File: docs/deployment/configuration.md (hunk 19 above)
Comment (fenekku) [reply docs/deployment/configuration.md]: Ah got it.

File: docs/deployment/configuration.md [hunk 27]
```diff
@@ -0,0 +1,84 @@
+# Configuration
+
+This section explains the current configuration options that are available for the different components deployed by these Helm charts.
```
Comment (fenekku) [inline docs/deployment/configuration.md]: See my previous comments about what "these Helm charts" refer to?

**Comment by fenekku** [comment]
I checked it again. Apart from the "these Helm charts" question, everything is good for me. (you can merge and we can go over this at some other point)
